crates/ty_python_semantic/src/semantic_index/place.rs

```diff
@@ -14,91 +15,303 @@ use crate::Db;
+
+    fn try_from(subscript: &ast::ExprSubscript) -> Result<Self, ()> {
+        let mut place = PlaceExpr::try_from(&*subscript.value)?;
+        match &*subscript.slice {
+            ast::Expr::NumberLiteral(number) if number.value.is_int() => {
+                place.sub_segments.push(PlaceExprSubSegment::IntSubscript(
+                    number.value.as_int().unwrap().clone(),
+                ));
+            }
+            ast::Expr::StringLiteral(string) => {
+                place
+                    .sub_segments
+                    .push(PlaceExprSubSegment::StringSubscript(
+                        string.value.to_string(),
+                    ));
+            }
+            _ => {
+                return Err(());
+            }
+        }
+        Ok(place)
+    }
+}
+
+impl TryFrom<ast::ExprSubscript> for PlaceExpr {
+    type Error = ();
+
+    fn try_from(subscript: ast::ExprSubscript) -> Result<Self, ()> {
+        PlaceExpr::try_from(&subscript)
+    }
+}
+
+impl TryFrom<&ast::Expr> for PlaceExpr {
+    type Error = ();
+
+    fn try_from(expr: &ast::Expr) -> Result<Self, ()> {
+        match expr {
+            ast::Expr::Name(name) => Ok(PlaceExpr::name(name.id.clone())),
+            ast::Expr::Attribute(attr) => PlaceExpr::try_from(attr),
+            ast::Expr::Subscript(subscript) => PlaceExpr::try_from(subscript),
+            _ => Err(()),
+        }
+    }
+}
+
+impl PlaceExpr {
+    pub(super) fn name(name: Name) -> Self {
         Self {
-            name,
-            flags: SymbolFlags::empty(),
+            root_name: name,
+            sub_segments: Vec::new(),
+            flags: PlaceFlags::empty(),
         }
     }
 
-    fn insert_flags(&mut self, flags: SymbolFlags) {
+    fn insert_flags(&mut self, flags: PlaceFlags) {
         self.flags.insert(flags);
     }
 
-    /// The symbol's name.
-    pub fn name(&self) -> &Name {
-        &self.name
+    pub(super) fn mark_instance_attribute(&mut self) {
+        self.flags.insert(PlaceFlags::IS_INSTANCE_ATTRIBUTE);
+    }
+
+    pub(crate) fn root_name(&self) -> &Name {
+        &self.root_name
+    }
+
+    pub(crate) fn sub_segments(&self) -> &[PlaceExprSubSegment] {
+        &self.sub_segments
+    }
+
+    pub(crate) fn as_name(&self) -> Option<&Name> {
+        if self.is_name() {
+            Some(&self.root_name)
+        } else {
+            None
+        }
+    }
+
+    /// Assumes that the place expression is a name.
+    #[track_caller]
+    pub(crate) fn expect_name(&self) -> &Name {
+        debug_assert_eq!(self.sub_segments, vec![]);
+        &self.root_name
+    }
+
+    /// Does the place expression have the form `self.{name}` (`self` is the first parameter of the method)?
+    pub(super) fn is_instance_attribute_named(&self, name: &str) -> bool {
+        self.flags.contains(PlaceFlags::IS_INSTANCE_ATTRIBUTE)
+            && self.sub_segments.len() == 1
+            && self.sub_segments[0].as_member().unwrap().as_str() == name
```

Comment:
Why is the `.unwrap()` safe here?

Reply:
If it is marked with `IS_INSTANCE_ATTRIBUTE`, it is an instance attribute and passes the `place_expr.is_member()` check.

https://github.com/mtshiba/ruff/blob/narrow_by_assignment/crates/ty_python_semantic/src/semantic_index/builder.rs#L1964-L1966